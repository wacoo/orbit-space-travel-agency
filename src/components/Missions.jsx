import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchMissions,
  joinMission,
  leaveMission,
} from '../redux/missions/missionsSlice';
import './Mission.css';

function Missions() {
  const missionData = useSelector((store) => store.missions) ?? {
    isLoading: false,
    missions: [],
    error: null,
    joinedMissions: [],
  };
  const {
    isLoading, missions, error, joinedMissions,
  } = missionData;
  const dispatch = useDispatch();

  function handleJoinMission(id) {
    dispatch(joinMission(id));
  }

  function handleLeaveMission(id) {
    dispatch(leaveMission(id));
  }

  useEffect(() => {
    if (missions.length === 0) {
      dispatch(fetchMissions());
    }
  }, [dispatch, missions.length]);

  if (isLoading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <>
      <table className="mission-table">
        <caption>Missions</caption>
        <thead>
          <tr>
            <th className="mission-cell">Mission</th>
            <th className="description-cell">Description</th>
            <th className="status-cell">Status</th>
            <th className="button-cell">&nbsp;</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <tr
              key={mission.id}
              className={
                joinedMissions.includes(mission.id) ? 'rows' : 'join-rows'
              }
            >
              <td className="mission-cell">{mission.name}</td>
              <td className="description-cell">
                {mission.description}
              </td>
              <td className="status-cell">
                {joinedMissions.includes(mission.id) ? (
                  <span className="leave-status">Active Member</span>
                ) : (
                  <span className="join-status">NOT A MEMBER</span>
                )}
              </td>
              <td className="button-cell">
                {joinedMissions.includes(mission.id) ? (
                  <button
                    type="button"
                    className="leave-button"
                    onClick={() => handleLeaveMission(mission.id)}
                  >
                    Leave Mission
                  </button>
                ) : (
                  <button
                    type="button"
                    className="join-button"
                    onClick={() => handleJoinMission(mission.id)}
                  >
                    Join Mission
                  </button>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
}

export default Missions;
