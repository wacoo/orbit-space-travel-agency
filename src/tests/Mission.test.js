import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { fireEvent, render, screen } from '@testing-library/react';
import Missions from '../components/Missions';

const mockStore = configureStore([thunk]);
let res = '';

const handleClick = () => {
  res = 'Clicked!';
};

describe('Test Missions Component', () => {
  test('renders Missions correctly', () => {
    const store = mockStore({
      missions: {
        isLoading: false,
        missions: [
          { id: 1, name: 'Mission 1', description: 'Description 1' },
          { id: 2, name: 'Mission 2', description: 'Description 2' },
        ],
        joinedMissions: [1],
      },
    });

    const { asFragment } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it('Tests Join Mission Button', () => {
    const store = mockStore({
      missions: {
        isLoading: false,
        missions: [
          { id: 1, name: 'Mission 1', description: 'Description 1' },
          { id: 2, name: 'Mission 2', description: 'Description 2' },
        ],
        joinedMissions: [1],
      },
    });
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const reserveButton = screen.getAllByText('Join Mission');
    reserveButton[0].onclick = handleClick();
    fireEvent.click(reserveButton[0]);
    expect(res).toBe('Clicked!');
  });
  it('Tests Leave Mission Button', () => {
    const store = mockStore({
      missions: {
        isLoading: false,
        missions: [
          { id: 1, name: 'Mission 1', description: 'Description 1' },
          { id: 2, name: 'Mission 2', description: 'Description 2' },
          { id: 3, name: 'Mission 3', description: 'Description 3' },
        ],
        joinedMissions: [1],
      },
    });
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const reserveButton = screen.getAllByText('Leave Mission');
    reserveButton[0].onclick = handleClick();
    fireEvent.click(reserveButton[0]);
    expect(res).toBe('Clicked!');
  });
});
