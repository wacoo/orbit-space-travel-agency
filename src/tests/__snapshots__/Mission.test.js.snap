// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Missions Component renders Missions correctly 1`] = `
<DocumentFragment>
  <table
    class="mission-table"
  >
    <thead>
      <tr>
        <th
          class="mission-cell"
        >
          Mission
        </th>
        <th
          class="description-cell"
        >
          Description
        </th>
        <th
          class="status-cell"
        >
          Status
        </th>
        <th
          class="button-cell"
        >
          Â 
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="rows"
      >
        <td
          class="mission-cell"
        >
          Mission 1
        </td>
        <td
          class="description-cell"
        >
          Description 1
        </td>
        <td
          class="status-cell"
        >
          <span
            class="leave-status"
          >
            Active Member
          </span>
        </td>
        <td
          class="button-cell"
        >
          <button
            class="leave-button"
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
      <tr
        class="join-rows"
      >
        <td
          class="mission-cell"
        >
          Mission 2
        </td>
        <td
          class="description-cell"
        >
          Description 2
        </td>
        <td
          class="status-cell"
        >
          <span
            class="join-status"
          >
            NOT A MEMBER
          </span>
        </td>
        <td
          class="button-cell"
        >
          <button
            class="join-button"
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
